### ISSUE TITLE
Anchoring: additions and adjustments

### ISSUE THREAD

**Issue by grahamperrin** [FreshPorts/freshports#196]
Please consider adding the following anchors. 

### `#add`

![image](https://user-images.githubusercontent.com/192271/105573769-d448a280-5d57-11eb-82c2-d6f86990e251.png)

* make it `#install` if you prefer, although (for subtle reasons, when providing IT support) I prefer `#add`
* anchored either to the _install_ line, or to the preceding white space. 

### `#message`

![image](https://user-images.githubusercontent.com/192271/105573883-85e7d380-5d58-11eb-978b-4df0cd5c9276.png)

* if you can add (and anchor to) a colourful divider, even better
* the significance of package messages may be easily overlooked by newcomers to FreeBSD.

### `#updating`

![image](https://user-images.githubusercontent.com/192271/105573916-c47d8e00-5d58-11eb-9bf1-221c3ac8c487.png)

* anchored to the colourful divider.

### `#history`

![image](https://user-images.githubusercontent.com/192271/105573949-f7278680-5d58-11eb-844e-74fa0963448d.png)

* I frequently want to refer to the commit history for a port. 

----

Thank you

**Comment by grahamperrin** [comment]
Please consider the following adjustments.

### `#requiredtobuild` …

![image](https://user-images.githubusercontent.com/192271/105574239-132c2780-5d5b-11eb-88dc-4bd1c35bef50.png)

* anchor to the subheading that precedes the first numbered item
* doing so can help with orientation, especially when a reader works with multiple tabs and/or windows.

The same for `#requiredtorun`, `#requiredlibraries`, `#requiredforBuild`, `#RequiredByBuild`, `#RequiredByRun` and `#requiredforRun`: 

* anchor to each listing's subheading. 

----

Thank you

**Comment by dlangille** [comment]
I've done these for now. I suspect some the others might be done.

Are you decent with HTML? We might need to update the [FAQ](https://github.com/FreshPorts/freshports/blob/master/www/faq.php)

- https://dev.freshports.org/www/firefox/#add
- https://dev.freshports.org/www/firefox/#message
- https://dev.freshports.org/www/firefox/#updating
- https://dev.freshports.org/www/firefox/#history

**Comment by grahamperrin** [comment]
Thanks! I'm OK with very basic HTML, if I don't do a PR for the FAQ today I'll (more likely) do so next weekend (with focus, and a clearer head). 

In the meantime …

----

Try anchoring `#add` to the white space above the text. It's subtly more legible. Readers are sometimes 'word blind' to text abutting the top of window. 

Re: <https://github.com/FreshPorts/freshports/issues/196#issuecomment-765893982>, `requiredforBuild` and `requiredforRun` are not yet anchored to their subheadings: 

![image](https://user-images.githubusercontent.com/192271/105628716-79d24380-5e36-11eb-94f5-0ed58be42c58.png)

You probably already know, some of the anchors currently appear more than once, for example: 

![image](https://user-images.githubusercontent.com/192271/105628876-81461c80-5e37-11eb-892c-e2b7c1c14082.png)

**Comment by dlangille** [comment]
I bet the duplicates are the result of prior work on anchors.

**Comment by dlangille** [comment]
> Try anchoring `#add` to the white space above the text. It's subtly more legible. Readers are sometimes 'word blind' to text abutting the top of window.

The challenge there is coding.

Consider `pkg-message`. There is already code like this. First, we have `USES`.

```
                if (($this->ShowEverything || $this->ShowUses) && $port->uses) {
                        $HTML .= '</dl><hr><dl><dt id="uses"><b>USES:</b></dt>' . "\n" . '<dd class="like-pre">';
                        $HTML .= $port->uses;
                        $HTML .= "</dd>\n</dl>\n<hr>\n<dl>";
                }

                if (($this->ShowEverything || $this->ShowPKGMessage) && $port->pkgmessage) {
                        $HTML .= $this->_pkgmessage($port);
                }
```

At present, the anchor code is handled within `_pkgmessage()` (not actually, but farther down in another function it calls).

The whitespace is a few

**Comment by grahamperrin** [comment]
Don't slave over it `:-)` it's the weekend. And so on. Thanks.

**Comment by dlangille** [comment]
The weekend is the best time for this.

**Comment by grahamperrin** [comment]
If it helps with any aspect of this: 

* https://addons.mozilla.org/addon/display-_anchors/
* https://addons.mozilla.org/addon/anchors-reveal/

**Display #Anchors** suits most cases. 

**Anchors Reveal** was, maybe still is, useful for edge cases where Display #Anchors could not display anchors (probably CSP-related … it's a _long_ time since I needed to make the distinction).

**Comment by dlangille** [comment]
I've amended the new anchors to not conflict with existing IDs.

- https://dev.freshports.org/www/firefox/#add
- https://dev.freshports.org/www/firefox/#message
- https://dev.freshports.org/www/firefox/#updating
- https://dev.freshports.org/www/firefox/#history
- https://dev.freshports.org/www/firefox/#requiredbuild
- https://dev.freshports.org/www/firefox/#requiredrun
- https://dev.freshports.org/www/firefox/#requiredlib
- https://dev.freshports.org/devel/electron9/#requiredfetch
- https://dev.freshports.org/devel/rubygem-logging/#requiredpatch
- https://dev.freshports.org/devel/rubygem-logging/#requiredextract

In short, slight changes, remove `to`, changed plural to singular.

**Comment by dlangille** [comment]
Try a different browser. This is Safari for https://dev.freshports.org/www/firefox/#requiredlib

![Screen Shot 2021-01-24 at 10 55 30 AM](https://user-images.githubusercontent.com/1626436/105635787-d92e5480-5e32-11eb-8ffa-4880ef09ce06.png)

**Comment by grahamperrin** [comment]
Thanks. Parallel testing in Falkon. 

Still dropping a little too low on these two: 

* https://dev.freshports.org/www/firefox/#requiredforBuild
* https://dev.freshports.org/www/firefox/#requiredforRun

Also I forgot to mention earlier, `#history` will (I think) be better as visualised: 

![image](https://user-images.githubusercontent.com/192271/105573949-f7278680-5d58-11eb-844e-74fa0963448d.png)

Consider https://dev.freshports.org/www/firefox/?page=2#history and so on; if you anchor a little higher, then the array of page numbers is immediately visible without scrolling up or down.

**Comment by grahamperrin** [comment]
FYI, an _Anchors Reveal_ view (full page screenshot – sorry, if it's distractingly tall, hide this comment):

![2021-01-24 17:35 full page](https://user-images.githubusercontent.com/192271/105638483-be77d200-5e6a-11eb-9a9f-e50e268aa7f3.png)

**Comment by dlangille** [comment]
> Still dropping a little too low on these two:
> 
> * https://dev.freshports.org/www/firefox/#requiredforBuild
> * https://dev.freshports.org/www/firefox/#requiredforRun

Please try these instead:

* https://dev.freshports.org/www/firefox/#requiredbuild
* https://dev.freshports.org/www/firefox/#requiredrun

The other values are not for anchor purposes.

> Consider https://dev.freshports.org/www/firefox/?page=2#history and so on; if you anchor a little higher, then the array of page numbers is immediately visible without scrolling up or down.

`#history` is now tied to `Number of commits found`

**Comment by grahamperrin** [comment]
> The other values are not for anchor purposes. Rather, not all what is show as anchors is meant for page positioning / linking.

OK, that explains it ☑

The other two (intended anchors) are fine. 

`#history` perfect now, I think. 

Thanks

**Comment by dlangille** [comment]
Any other work to do in this area? I'm watching terrible movies and coding.

**Comment by grahamperrin** [comment]
If the movies are too terrible to watch, you could jazz up the arrays of history section page numbers: 

![image](https://user-images.githubusercontent.com/192271/105639340-ae162600-5e6f-11eb-9535-d3bd1a2e5d43.png)

– for each linked item in the array to have a `#history` tail: 

* https://dev.freshports.org/www/firefox/?page=2#history

– and so on. 

I'm eating a Pot Noodle and on the edge of my seat not with excitement, but because one of my cats deems the rear two-thirds of the seat to be the warmest and most stimulating place in the building.

**Comment by dlangille** [comment]
That took more time finding than changing:

```
-                       'fileName'              => '?page=%d',
+                       'fileName'              => '?page=%d#history',
```

**Comment by dlangille** [comment]
Now in test: https://test.freshports.org/www/firefox/#requiredbuild

**Comment by dlangille** [comment]
Staging: https://stage.freshports.org/www/firefox/#requiredbuild

**Comment by dlangille** [comment]
And in production: https://www.freshports.org/www/firefox/#requiredbuild

**Comment by dlangille** [comment]
@grahamperrin FAQ completed: https://dev.freshports.org/faq.php

**Comment by grahamperrin** [comment]
Ah I just started on it! Thanks

**Comment by dlangille** [comment]
@grahamperrin sorry, I was watching boring tv.  But thank you. if you have something nicer, please advise.

**Comment by grahamperrin** [comment]
Big thanks! 

FYI https://github.com/FreshPorts/freshports/commit/9572b879e71964ad18a674a9130373c122fe7b5b#r46329537

… then https://github.com/FreshPorts/freshports/pull/200 is primarily for linking to _anchored answers_ to frequently asked questions.
